### PR TITLE
Document testing with `@EmbeddedKafka`

### PR DESCRIPTION
**Pull request by artembilan**
Provide an auto-configuration when an `@EmbeddedKafka` is used
in the tests.

The implementation just re-maps an
`EmbeddedKafkaBroker.getBrokersAsString()` into the environment to
let end-users do not do that manually via some additional system
or application properties manipulation

In addition, the `KafkaAutoConfiguration` now consults environment
for that new property to override provided one from the
application properties

### CODE REVIEW

File: spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/kafka/KafkaStreamsAnnotationDrivenConfiguration.java [hunk 1]
```diff
@@ -72,6 +79,14 @@ public KafkaStreamsFactoryBeanConfigurer kafkaStreamsFactoryBeanConfigurer(
 		return new KafkaStreamsFactoryBeanConfigurer(this.properties, factoryBean);
 	}
 
+	private Map<String, Object> applyEmbeddedBrokersIfAny(Map<String, Object> properties) {
+		String embeddedBrokerAddresses = this.environment.getProperty("spring.embedded.kafka.brokers");
```
[the remark this answers is not in the document]
Comment (artembilan) [reply spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/kafka/KafkaStreamsAnnotationDrivenConfiguration.java]: That is the name of the system property provided by the `EmbeddedKafkaBroker`: https://docs.spring.io/spring-kafka/docs/2.2.3.RELEASE/reference/html/_reference.html#_junit

But I see your point and will change it according Spring Boot nature, leave the original system property as is do not break something else.
Although essentially they are going to have the same value anyway.

**Comment by artembilan** [comment]
I somehow missed the fact that `spring-boot-test-autoconfigure` is there :shame:.
And yes I agree that this really belongs there and there is nothing with embedded mode like we have with MongoDB or JDBC.
Only the trigger for this feature is really a test marked with the `@EmbeddedKafka`.

**Comment by artembilan** [comment]
The concern about embedded brokers property knowledge is minor since that's only string literal.

My problem, though, that I haven't figured out how to remap that embedded brokers property into the `spring.kafka.bootstrapServers`  and have it conditional via specific auto-configuration class.
I just borrowed the current logic from the MongoDB auto-configuration and its Embedded mode.

Thanks for your feedback, @wilkinsona ! 
Will incorporate it today or tomorrow...

File: spring-boot-project/spring-boot-autoconfigure/src/test/java/org/springframework/boot/autoconfigure/kafka/KafkaAutoConfigurationIntegrationTests.java [hunk 2]
```diff
@@ -45,75 +50,67 @@
  *
  * @author Gary Russell
  * @author Stephane Nicoll
+ * @author Artem Bilan
  */
+@RunWith(SpringRunner.class)
+@SpringBootTest(
```
[the remark this answers is not in the document]
Comment (artembilan) [reply spring-boot-project/spring-boot-autoconfigure/src/test/java/org/springframework/boot/autoconfigure/kafka/KafkaAutoConfigurationIntegrationTests.java]: The main point for the feature I want to promote is really about an `@EmbeddedKafka` on the test class which can be triggered only by the `@RunWith(SpringRunner.class)`.
So, `ApplicationContextRunner` is not going to help here.

Hope after moving this to the `spring-boot-test-autoconfigure` it is not going to look so aggressive...

**Comment by artembilan** [comment]
I have moved `EmbeddedKafkaAutoConfiguration` into the `spring-boot-test-autoconfigure`.

Please, pay attention that it is still simple holder for the regular `EnableAutoConfiguration`. There is no something like `@KafkaTest` slice, like we have for many other features.

The `EmbeddedKafkaAutoConfiguration` fully depends on the presence of the `@EmbeddedKafka` annotation on the test.
And from here that `EmbeddedKafkaAutoConfiguration` has a condition like this: `@ConditionalOnBean(EmbeddedKafkaBroker.class)`.

So far I don't see reason in extra functionality like `@KafkaTest` and respective `@AutoConfigureKafka` and so on supporting classes.
For me a well-known `@EmbeddedKafka` is fully enough to achieve a goal.

Thanks

**Comment by wilkinsona** [comment]
I've rebased and squashed the proposed changes and pushed them [here](https://github.com/wilkinsona/spring-boot/tree/gh-15834). Having done that, on closer inspection the proposal doesn't feel quite right for a couple of reasons:

1. It introduces a conceptual cycle between `spring-boot-autoconfigure` and `spring-boot-test-autoconfigure`. The latter has a build dependency on the former and the former depends on the later via knowledge of the `spring.kafka.embedded.brokers` property.
2. It adds a `spring.factories` `EnableAutoConfiguration` entry for `EmbeddedKafkaAutoConfiguration` in `spring-boot-test-autoconfigure`, the first such entry in the module.

I'm not entirely sure what to do about either of these. For 1 I wonder if we should set the `spring.kafka.bootstrap-servers` property directly. For 2 I wonder if we should have an `@AutoConfigureEmbeddedKafka` annotation.

Flagging for team attention so that we can hopefully figure out what to do.

**Comment by artembilan** [comment]
@wilkinsona ,

thank you for feedback!
I need to refresh it in my mind. The last change to this PR has been done almost half-year ago 😢 
I'll be back with some answers later...

**Comment by artembilan** [comment]
Well, I'm OK to just populate an additional `spring.kafka.bootstrap-servers` system property in the Spring Kafka and forget this doubtful solution with re-mapping.

@garyrussell , WDYT?

**Comment by garyrussell** [comment]
I guess that's ok (as long as it is reset after) in case they have other tests that run against a real broker.

Of course, it introduces the same conceptual cycle, but now between spring-boot and spring-kafka (spring-kafka having knowledge of boot's property).

TBH, I don't think it's too painful to have them set it manually, like we suggest in the [documentation](https://docs.spring.io/spring-kafka/reference/html/#kafka-testing-embeddedkafka-annotation). Although, I suppose, a `TestPropertySource` would be better.

**Comment by artembilan** [comment]
Well, the point of this change because somehow people are missing to set such a property manually. I came to this solution because I had answered the same problem in different place in one day, as far as I recall.
Such a simple auto-configuration magic does the trick for us without any system properties manipulations and without changing an original `spring.kafka.bootstrap-servers` property.

I understand that something like a new `@AutoConfigureEmbeddedKafka` would do the trick for us as well, but who know how people are going to forget it as an option for Spring Boot. Plus I believe they will come back to us with questions what is the point of this one since we have that nice `@EmbeddedKafka` one already. Why just don't make an existing solution working automatically in Spring Boot without any further action? We have a configuration on convention policy here, so, can't we just close our eyes and have that plain string dependency for one extra property?
I believe we even can move that `EmbeddedKafkaAutoConfiguration` back to the `spring-boot-autoconfigure`. I would say the `spring-kafka-test` plays in Spring Boot exactly the same role as `de.flapdoodle.embed.mongo`. Looks like that one doesn't bring any questions, therefore we can simply treat an embedded Kafka as a feature not just testing utility.

**Comment by wilkinsona** [comment]
> Well, I'm OK to just populate an additional spring.kafka.bootstrap-servers system property in the Spring Kafka and forget this doubtful solution with re-mapping.

Sorry I wasn't clear. That's not what I meant. I dislike that conceptual cycle even more than one between `spring-boot-autoconfigure` and `spring-boot-test-autoconfigure`. I was wondering if Boot could map your existing system property to `spring.kafka.bootstrap-servers` entry in the environment to avoid `spring-boot-autoconfigure` having to know about `spring.kafka.embedded.brokers` that's currently being set by `EmbeddedKafkaAutoConfiguration`.

**Comment by artembilan** [comment]
Yes, I think we can't do that because Spring Boot freezes properties after loading them into environment. We just can't override it from the auto-configuration. Therefore a new one to rely during beans initialization afterwards.
Am I wrong?

**Comment by wilkinsona** [comment]
We control the ordering so we can guarantee that the embedded Kafka auto-configuration runs before the main Kafka auto-configuration. This would give the embedded auto-configuration a chance to create a dependency relationship between beans that ensures that embedded Kafka is running and that a property has been set in the environment before the main Kafka auto-configuration triggering the binding of the property and tries to connect to Kafka.

**Comment by artembilan** [comment]
@wilkinsona ,

Thank you for your patience and valuable feedback.
So, I fixed my changes according your request and now the `EmbeddedKafkaAutoConfiguration` really populates a `spring.kafka.bootstrap-servers` directly.
Looks like `Environment` is not so frozen in case of test environment. Or I just misunderstood something about that in the past...
Anyway it is much cleaner now and there is no any changes in the `spring-boot-autoconfigure` at all!

Enjoy! 😄

**Comment by wilkinsona** [comment]
Thanks for the updates, Artem. We'll also need some configuration to make sure that `EmbeddedKafkaAutoConfiguration` creates the property source before the client is created. As exactly what is involved here becomes clearer, the following point from @garyrussell is becoming more and more persuasive:

> TBH, I don't think it's too painful to have them set it manually, like we suggest in the documentation. 

That minimal amount of pain is starting to feel like a better option than the complexity of trying to do it automatically, particularly as we'd currently end up with `EmbeddedKafkaAutoConfiguration` being an odd-one-out in the `spring-boot-test-autoconfigure` module.

**Comment by wilkinsona** [comment]
> TBH, I don't think it's too painful to have them set it manually

Let's leave things as they are for 2.2 and see what feedback we get from users about how painful it is to do it manually.

**Comment by artembilan** [comment]
OK, Andy.

If I am in the minority, then let me to convert this PR into Doc changes and mention such a manual step to apply when you use Embedded Kafka in your Spring Boot tests!

**Comment by artembilan** [comment]
@wilkinsona ,

take a look, please, to this minimal requirements from all those ours discussions.

Thank you!

**Comment by garyrussell** [comment]
LGTM

**Comment by garyrussell** [comment]
However, please see: https://github.com/spring-projects/spring-kafka/pull/1173 - I think that's cleaner.

**Comment by artembilan** [comment]
OK. Since I have just merged that PR, I also can add such an option into this Doc change.

Does it make sense?

**Comment by garyrussell** [comment]
Yes; this should reflect that change.

**Comment by wilkinsona** [comment]
Thanks, Artem. I'm glad we managed to settle on something in the end. The docs updates are now in master.

**Pull request closed without merging**